wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/convert/DefaultEditorProvisioner.java

```diff
@@ -99,20 +104,97 @@ public void apply(DevfileImpl devfile, FileContentProvider contentProvider)
     }
   }
 
+  /**
+   * Provision the default editor plugins and add them to the the Devfile's component list
+   *
+   * @param components The set of components currently present in the Devfile
+   * @param contentProvider content provider for plugin references retrieval
+   * @throws DevfileException
+   */
   private void provisionDefaultPlugins(
       List<ComponentImpl> components, FileContentProvider contentProvider) throws DevfileException {
     Map<String, String> missingPluginsIdToRef = new HashMap<>(defaultPluginsToRefs);
+    removeAlreadyAddedPlugins(components, contentProvider, missingPluginsIdToRef);
+    addMissingPlugins(components, contentProvider, missingPluginsIdToRef);
+  }
 
+  /**
+   * Checks if any of the Devfile's components are also in the list of missing default plugins, and
+   * removes them.
+   *
+   * @param components - The list of Devfile components
+   * @param contentProvider - The content provider to retrieve YAML
+   * @param missingPluginsIdToRef - The list of default plugins that are not currently in the list
+   *     of Devfile components
+   */
+  private void removeAlreadyAddedPlugins(
+      List<ComponentImpl> components,
+      FileContentProvider contentProvider,
+      Map<String, String> missingPluginsIdToRef)
+      throws DevfileException {
     for (ComponentImpl t : components) {
       if (PLUGIN_COMPONENT_TYPE.equals(t.getType())) {
         String pluginPublisherAndName = getPluginPublisherAndName(t, contentProvider);
         missingPluginsIdToRef.remove(pluginPublisherAndName);
       }
     }
+  }
 
-    missingPluginsIdToRef
-        .values()
-        .forEach(pluginRef -> components.add(new ComponentImpl(PLUGIN_COMPONENT_TYPE, pluginRef)));
+  /**
+   * Tries to add default plugins to the Devfile components. Each plugin is initially parsed by
+   * plugin ref. If the plugin does not have a reference, it is added to the component list, and its
+   * plugin ID will be used to resolve it. If it has a reference, the Plugin is evaluated, so that
+   * its meta.yaml can be retrieved. From the meta.yaml, the new Component's ID and reference are
+   * properly set, and the Component is added to the list.
+   *
+   * @param components - The list of Devfile components
+   * @param contentProvider - The content provider to retrieve YAML
+   * @param missingPluginsIdToRef - The list of default plugins that are not currently in the list
+   *     of devfile components
+   * @throws DevfileException
+   */
+  private void addMissingPlugins(
+      List<ComponentImpl> components,
```

Comment:
can we name it `devfileComponents` so it's clear what it is ?

Reply:
renamed to `devfileComponents`